File: internal/webserver/instances.go
package webserver

import (
	"encoding/json"
	"net/http"
	"path/filepath"
	"slices"
	"strconv"
	"time"

	"github.com/RomkaLTU/trau/internal/folderrepo"
	"github.com/RomkaLTU/trau/internal/logger"
	"github.com/RomkaLTU/trau/internal/registry"
)

// Instance is a live loop as the hub sees it: the registry record plus the
// session state the loop reports through its heartbeat. The hub echoes that
// state verbatim and never derives activity from run artifacts.
type Instance struct {
	PID          int    `json:"pid"`
	Repo         string `json:"repo"`
	RepoRoot     string `json:"repo_root"`
	RunsDir      string `json:"runs_dir"`
	StartedAt    string `json:"started_at"`
	SessionState string `json:"session_state"`
	Ticket       string `json:"ticket,omitempty"`
	Phase        string `json:"phase,omitempty"`
	Activity     string `json:"activity,omitempty"`
	Detail       string `json:"detail,omitempty"`
	StateSince   string `json:"state_since,omitempty"`
}

// RepoView is a repo the hub knows about, flagged with whether a loop is
// currently running in it and whether the hub may start one there. Repos linger
// here after their loop exits so their runs stay browsable; an unallowed repo is
// observe-only. Registered marks a repo whose startability comes from a web
// registration rather than the SERVE_WORKSPACE seed, so the UI offers unregister
// only where it applies. Seeded marks the config-owned grant no removal can take
// back, which a registration can overlap. Kind tells a Folder repo from an
// ordinary one and ChildRepos counts the Child repos behind it. Freshness carries
// the issue-store sync state and is attached only on the repos API, where the
// background sync surfaces it.
type RepoView struct {
	registry.Repo
	Live       bool           `json:"live"`
	Allowed    bool           `json:"allowed"`
	Registered bool           `json:"registered"`
	Seeded     bool           `json:"seeded"`
	Kind       string         `json:"kind"`
	ChildRepos int            `json:"child_repos,omitempty"`
	Freshness  *RepoFreshness `json:"freshness,omitempty"`
}

const (
	repoKindRepo   = "repo"
	repoKindFolder = "folder"
)

// withKind fills in the folder facts, derived on every read rather than stored:
// an ordinary repo costs the one stat the scan short-circuits on, and only a real
// Folder repo pays the directory read. The instances resource polls every few
// seconds, and that scan is the price of a count that is always right and a
// registration store that never has to be told a child was cloned or removed. A
// directory that cannot be read reads as an ordinary repo rather than failing the
// request.
func (v RepoView) withKind() RepoView {
	v.Kind = repoKindRepo
	if census := folderrepo.Scan(v.Root); census.IsFolder {
		v.Kind, v.ChildRepos = repoKindFolder, len(census.Children)
	}
	return v
}

// RepoFreshness is a repo's issue-store freshness: its derived health state, when
// it last synced from the tracker, whether a background sync is running right now,
// the error from the last failed attempt and what it takes to clear (both empty
// once a sync succeeds), the counts the last good sync wrote, and how many issues
// the store now holds. On the repos API it always carries a State so the Instances
// page renders a designed state; the backlog attaches only the sync fields, leaving
// State and IssueCount unset.
type RepoFreshness struct {
	State         RepoHealthState `json:"state,omitempty"`
	LastSyncedAt  string          `json:"last_synced_at,omitempty"`
	Syncing       bool            `json:"syncing"`
	LastError     string          `json:"last_error,omitempty"`
	LastErrorKind string          `json:"last_error_kind,omitempty"`
	LastIssues    int             `json:"last_issues,omitempty"`
	LastComments  int             `json:"last_comments,omitempty"`
	IssueCount    int             `json:"issue_count,omitempty"`
}

// InstancesResponse is the /api/v1/instances resource: the live loops and every
// repo the hub has ever seen a loop run in. TakeoverSupported reports whether
// this hub's platform can open a terminal takeover (ADR 0018), so the web can
// hide the action instead of offering a guaranteed 501.
type InstancesResponse struct {
	Instances         []Instance `json:"instances"`
	Repos             []RepoView `json:"repos"`
	TakeoverSupported bool       `json:"takeover_supported"`
}

// instanceHeartbeatBody is a loop's reported presence on a register or heartbeat.
// It mirrors hubclient.InstanceHeartbeat: the hub keys presence by the {pid} path
// segment and stamps its own last-seen, so the body carries only what the loop
// reports.
type instanceHeartbeatBody struct {
	RepoRoot     string    `json:"repo_root"`
	RunsDir      string    `json:"runs_dir"`
	StartedAt    time.Time `json:"started_at"`
	SessionState string    `json:"session_state"`
	Ticket       string    `json:"ticket,omitempty"`
	Phase        string    `json:"phase,omitempty"`
	Activity     string    `json:"activity,omitempty"`
	Detail       string    `json:"detail,omitempty"`
	StateSince   time.Time `json:"state_since,omitzero"`
}

func (s *Server) handleInstances(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.Header().Set("Allow", http.MethodGet)
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
		return
	}
	s.listInstances(w, r)
}

// handleInstance is a loop's presence seam (ADR 0008 §7): the loop PUTs its
// heartbeat — on start, on every session-state change, and on a timer — keyed by
// its PID, and DELETEs it on clean exit. The hub echoes the reported state and
// reaps a dead PID via signal 0, so a crashed loop that never DELETEs still ages
// out. Presence is best-effort on the loop side; the hub answers plainly. A PID
// the store has never seen is a loop starting and a DELETE is one settling, which
// is where team sync folds the teammates' records in and publishes this machine's.
func (s *Server) handleInstance(w http.ResponseWriter, r *http.Request) {
	pid, err := strconv.Atoi(r.PathValue("pid"))
	if err != nil || pid <= 0 {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid pid"})
		return
	}
	instances := s.stores.Instances()

	switch r.Method {
	case http.MethodPut:
		var req instanceHeartbeatBody
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
			return
		}
		entry := registry.Entry{
			PID:          pid,
			RepoRoot:     req.RepoRoot,
			RunsDir:      req.RunsDir,
			StartedAt:    req.StartedAt,
			Heartbeat:    time.Now(),
			SessionState: req.SessionState,
			Ticket:       req.Ticket,
			Phase:        req.Phase,
			Activity:     req.Activity,
			Detail:       req.Detail,
			StateSince:   req.StateSince,
		}
		_, known, err := instances.Get(pid)
		if err != nil {
			logger.Verbosef("instances get %d: %v", pid, err)
		}
		if err := instances.Upsert(entry); err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		if !known {
			s.team.kick(entry.RepoRoot)
		}
		writeJSON(w, http.StatusOK, map[string]any{"status": "ok", "pid": pid})
	case http.MethodDelete:
		entry, _, err := instances.Get(pid)
		if err != nil {
			logger.Verbosef("instances get %d: %v", pid, err)
		}
		if err := instances.Remove(pid); err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		s.team.kick(entry.RepoRoot)
		writeJSON(w, http.StatusOK, map[string]any{"status": "removed", "pid": pid})
	default:
		w.Header().Set("Allow", "PUT, DELETE")
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
	}
}

// liveInstances is the hub's read of the live loops on this machine, from the
// authoritative presence store: the entries whose PID still passes signal 0, with
// the dead ones reaped. A store error reads as no live loops, keeping every
// consumer's fail-safe (nothing live) the same as the file-era empty glob.
func (s *Server) liveInstances() []registry.Entry {
	entries, err := s.stores.Instances().Live()
	if err != nil {
		logger.Verbosef("instances live: %v", err)
		return nil
	}
	return entries
}

// repoIsLive reports whether any loop is registered in root right now, idle
// dashboards included, matching what the repos list flags as live. The presence
// sweep remembers every live loop's repo, so removing one would not stick. An
// entry's root and the stored row's can spell the same directory differently, and
// the refusal has to fire on either spelling.
func (s *Server) repoIsLive(root string) bool {
	canonical := registry.CanonicalRoot(root)
	return slices.ContainsFunc(s.liveInstances(), func(e registry.Entry) bool {
		return registry.CanonicalRoot(e.RepoRoot) == canonical
	})
}

// folderCollisionError refuses work in a repo whose enclosing Folder repo, or one
// of whose Child repos, already has a loop live in it. It names both the repo and
// the ticket its loop is on, so the board states the reason before anything spawns.
type folderCollisionError struct{ entry registry.Entry }

func (e folderCollisionError) Error() string {
	return folderrepo.CollisionReason(e.entry.Ticket, e.entry.RepoRoot)
}

// folderCollision names the live loop a run in root would share a working tree
// with: one in a Child repo when root is a Folder repo, or one in the Folder repo
// that holds root. Every live entry blocks, a takeover included — a terminal
// session holds that working tree just as firmly as a loop does.
func (s *Server) folderCollision(root string) (folderCollisionError, bool) {
	for _, e := range s.liveInstances() {
		if folderrepo.Collides(root, e.RepoRoot) {
			return folderCollisionError{entry: e}, true
		}
	}
	return folderCollisionError{}, false
}

// hasBusyInstance reports whether a live instance in root is past idle — a run in
// flight, held WIP, or a takeover terminal owning the working tree (ADR 0018). An
// idle instance is an open dashboard rather than a run, and a legacy entry with no
// state counts as busy. An empty root asks across every repo.
func (s *Server) hasBusyInstance(root string) bool {
	for _, e := range s.liveInstances() {
		if root != "" && e.RepoRoot != root {
			continue
		}
		if e.SessionState != registry.StateIdle {
			return true
		}
	}
	return false
}

func (s *Server) listInstances(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, InstancesResponse{
		Instances:         s.instanceViews(),
		Repos:             s.repoViews(),
		TakeoverSupported: s.takeoverSupported(),
	})
}

// instanceViews maps the live presence entries onto the JSON instance rows the
// Instances page and the MCP tool share.
func (s *Server) instanceViews() []Instance {
	entries := s.liveInstances()
	instances := make([]Instance, 0, len(entries))
	for _, e := range entries {
		inst := Instance{
			PID:       e.PID,
			Repo:      filepath.Base(e.RepoRoot),
			RepoRoot:  e.RepoRoot,
			RunsDir:   e.RunsDir,
			StartedAt: e.StartedAt.UTC().Format(time.RFC3339),
		}
		if e.SessionState == "" {
			inst.SessionState = "unknown"
		} else {
			inst.SessionState = e.SessionState
			inst.Ticket = e.Ticket
			inst.Phase = e.Phase
			inst.Activity = e.Activity
			inst.Detail = e.Detail
			if !e.StateSince.IsZero() {
				inst.StateSince = e.StateSince.UTC().Format(time.RFC3339)
			}
		}
		instances = append(instances, inst)
	}
	return instances
}

File: web/src/lib/instances.ts
import { queryOptions, useQuery } from "@tanstack/react-query";

import type { RunState } from "@/components/trau/status-pill";
import { apiFetch } from "./api";

export interface Instance {
  pid: number;
  repo: string;
  repo_root: string;
  runs_dir: string;
  started_at: string;
  session_state: string;
  ticket?: string;
  phase?: string;
  activity?: string;
  detail?: string;
  state_since?: string;
}

// RepoHealthState mirrors the hub's derived health. A recorded error over an
// earlier good sync is degraded — the store still holds that pull's issues — while
// sync-failed is an error with no local data behind it.
export type RepoHealthState =
  | "ready"
  | "unconfigured"
  | "degraded"
  | "sync-failed"
  | "never-synced"
  | "syncing";

// SyncErrorKind is what the hub decided the last sync error takes to clear: a
// rate limit or a transport failure retries itself, while a config failure waits
// on the repo's settings or credentials. Empty when no error stands.
export type SyncErrorKind = "" | "config" | "rate-limit" | "transient";

// RepoFreshness is a repo's issue-store sync state: when it last synced from the
// tracker, whether a background sync is running right now, the last sync error,
// and the counts the last good sync wrote. Absent for a repo that has never
// synced and is not syncing. The repos API always carries a state and issue_count;
// the backlog attaches only the sync fields.
export interface RepoFreshness {
  state?: RepoHealthState;
  last_synced_at?: string;
  syncing: boolean;
  last_error?: string;
  last_error_kind?: SyncErrorKind;
  last_issues?: number;
  last_comments?: number;
  issue_count?: number;
}

// kind tells a Folder repo — a registered folder whose git children are the
// repositories a run ships to — from an ordinary one; child_repos counts those
// children and is set only for a folder.
export interface RepoView {
  name: string;
  root: string;
  runs_dir: string;
  live: boolean;
  allowed: boolean;
  registered: boolean;
  seeded: boolean;
  kind: "repo" | "folder";
  child_repos?: number;
  freshness?: RepoFreshness;
}

// RepoHealth is one repo's /health resource: the derived state plus the sync
// facts behind it, so a gate can poll a single repo instead of the whole list.
// provider is the tracker the state came from — empty or "internal" means there
// is nothing to pull.
export interface RepoHealth {
  repo: string;
  provider: string;
  state: RepoHealthState;
  last_synced_at: string;
  last_error: string;
  last_error_kind: SyncErrorKind;
  issue_count: number;
}

// externalTracker reports whether a repo's provider is one there is anything to
// pull from. It is the rule every sync affordance gates on: an empty provider is
// a repo with no tracker configured, the internal tracker lives in the hub
// itself, and the sync endpoint refuses both.
export function externalTracker(provider?: string): boolean {
  return !!provider && provider !== "internal";
}

export interface InstancesResponse {
  instances: Instance[];
  repos: RepoView[];
  takeover_supported?: boolean;
}

// Only a store read error drops the freshness the repos API otherwise always
// sends; that repo reads as unconfigured rather than claiming health.
export function repoHealth(repo: RepoView): RepoHealthState {
  return repo.freshness?.state ?? "unconfigured";
}

export function anySyncing(repos: readonly RepoView[]): boolean {
  return repos.some((repo) => repoHealth(repo) === "syncing");
}

export function healthPill(state: RepoHealthState): {
  state: RunState;
  label: string;
} {
  switch (state) {
    case "ready":
      return { state: "success", label: "ready" };
    case "syncing":
      return { state: "active", label: "syncing" };
    case "sync-failed":
      return { state: "fail", label: "sync failing" };
    case "degraded":
      return { state: "warn", label: "sync degraded" };
    case "never-synced":
      return { state: "warn", label: "never synced" };
    case "unconfigured":
      return { state: "warn", label: "not configured" };
  }
}

// healthBlocks reports whether a state leaves a repo-scoped page with nothing to
// show: nothing is configured to fetch, or every sync so far has failed. Degraded
// is deliberately absent — the store holds the last good pull, so the page renders
// it under a notice rather than behind an overlay.
export function healthBlocks(state: RepoHealthState): boolean {
  return state === "unconfigured" || state === "sync-failed";
}

async function fetchInstances(): Promise<InstancesResponse> {
  const res = await apiFetch("/api/v1/instances");
  if (!res.ok) {
    throw new Error(`instances request failed: ${res.status}`);
  }
  return res.json();
}

export const instancesQueryOptions = queryOptions({
  queryKey: ["instances"],
  queryFn: fetchInstances,
  refetchInterval: 3000,
  // Keep polling in a backgrounded tab so the live tab title stays current while
  // a user watches a run from another tab.
  refetchIntervalInBackground: true,
});

// TAKEOVER_BLOCKED is what every control a live takeover gates says, both as a
// tooltip and as the rendering of the hub's 409 (ADR 0018).
export const TAKEOVER_BLOCKED =
  "Taken over in a terminal — close it to hand the repo back";

export function repoTakenOver(
  instances: readonly Instance[],
  repo: string,
): boolean {
  return instances.some(
    (i) => i.repo === repo && i.session_state === "takeover",
  );
}

// useRepoTakenOver rides the shared instances poll, so a gated control re-enables
// on its own once the takeover deregisters.
export function useRepoTakenOver(repo: string): boolean {
  const { data } = useQuery(instancesQueryOptions);
  return data ? repoTakenOver(data.instances, repo) : false;
}

export function repoRoot(repos: readonly RepoView[], repo: string): string {
  return repos.find((r) => r.name === repo)?.root ?? "";
}

// useRepoRoot rides the shared instances poll, so a surface that needs the repo's
// absolute path gets it without a fetch of its own.
export function useRepoRoot(repo: string): string {
  const { data } = useQuery(instancesQueryOptions);
  return data ? repoRoot(data.repos, repo) : "";
}

async function fetchRepoHealth(repo: string): Promise<RepoHealth> {
  const res = await apiFetch(
    `/api/v1/repos/${encodeURIComponent(repo)}/health`,
  );
  if (!res.ok) {
    throw new Error(`repo health request failed: ${res.status}`);
  }
  return res.json();
}

// How often the health of a repo is re-read: rarely while it sits still, every
// few seconds while a pull is in flight so the state — and the board waiting on
// it — follows that pull within seconds of it landing.
const HEALTH_POLL_MS = 30_000;
const HEALTH_SYNCING_POLL_MS = 3_000;

// Keyed by repo so every gate on a page shares one fetch rather than one per
// section. It polls because the pages behind it stay open for hours: without an
// interval a background sync that started failing would go unnoticed until the
// next focus or remount.
export const repoHealthQueryOptions = (repo: string) =>
  queryOptions({
    queryKey: ["repo-health", repo],
    queryFn: () => fetchRepoHealth(repo),
    enabled: repo !== "",
    staleTime: 15_000,
    refetchInterval: (query) =>
      query.state.data?.state === "syncing"
        ? HEALTH_SYNCING_POLL_MS
        : HEALTH_POLL_MS,
  });

async function errorMessage(res: Response, fallback: string): Promise<string> {
  const detail = (await res.json().catch(() => null)) as {
    error?: string;
  } | null;
  return detail?.error ?? `${fallback}: ${res.status}`;
}

export interface StopResult {
  status: string;
  signal: string;
}

export async function stopInstance(pid: number): Promise<StopResult> {
  const res = await apiFetch(`/api/v1/instances/${pid}/stop`, {
    method: "POST",
  });
  if (!res.ok) {
    throw new Error(await errorMessage(res, "stop failed"));
  }
  return res.json();
}

export class TakeoverError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "TakeoverError";
    this.status = status;
  }
}

export interface TakeoverResult {
  stopped: boolean;
  opened: boolean;
}

// takeoverRun hands a run to a terminal (ADR 0018). On a live run the endpoint
// blocks through the graceful stop — up to 60s — so the fetch timeout sits past
// that bound instead of aborting mid-stop.
export async function takeoverRun(
  repo: string,
  ticket: string,
): Promise<TakeoverResult> {
  const res = await apiFetch(
    `/api/v1/repos/${encodeURIComponent(repo)}/runs/${encodeURIComponent(
      ticket,
    )}/takeover`,
    { method: "POST", signal: AbortSignal.timeout(75_000) },
  );
  if (!res.ok) {
    throw new TakeoverError(
      await errorMessage(res, "takeover failed"),
      res.status,
    );
  }
  return res.json();
}

export async function registerRepo(path: string): Promise<RepoView> {
  const res = await apiFetch("/api/v1/repos", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ path }),
  });
  if (!res.ok) {
    throw new Error(await errorMessage(res, "register failed"));
  }
  return res.json();
}

// unregisterRepo and forgetRepo both address the repo by ident — its name, or its
// root path, which is the only way to name one of two repos sharing a basename.
export async function unregisterRepo(ident: string): Promise<RepoView> {
  const res = await apiFetch(`/api/v1/repos/${encodeURIComponent(ident)}`, {
    method: "DELETE",
  });
  if (!res.ok) {
    throw new Error(await errorMessage(res, "unregister failed"));
  }
  return res.json();
}

// forgetRepo drops the repo off the hub's list entirely, the one removal that
// also reaches a repo the hub merely observed a loop run in.
export async function forgetRepo(ident: string): Promise<RepoView> {
  const res = await apiFetch(
    `/api/v1/repos/${encodeURIComponent(ident)}?forget=1`,
    { method: "DELETE" },
  );
  if (!res.ok) {
    throw new Error(await errorMessage(res, "remove failed"));
  }
  return res.json();
}

// removeBlocked reports why the hub would refuse to remove this repo, so the row
// says so up front instead of handing back a toast. It reads the seeded flag the
// hub sends rather than inferring the grant from a missing registration, which a
// repo that is both seeded and registered would read the wrong way round. It weighs
// the two refusals in the hub's order, config grant before live loop, so a repo that
// is both is refused here for the same reason the hub gives. null means the removal
// goes through.
export function removeBlocked(repo: RepoView): string | null {
  if (repo.seeded) {
    return "Granted by SERVE_WORKSPACE — drop its root from the config instead";
  }
  if (repo.live) {
    return "A loop is live here — stop it before removing the repo";
  }
  return null;
}

export interface SyncResponse {
  repo: string;
  provider: string;
  issues: number;
  comments: number;
  removed: number;
  syncedAt: string;
}

// SyncResult is how a sync request settled: the counts a pull of our own wrote,
// or the hub's 202 marker that it coalesced into a pull already in flight — no
// counts to show, and nothing that failed.
export type SyncResult =
  | { status: "pulled"; response: SyncResponse }
  | { status: "syncing" };

export function pullCounts(res: SyncResponse): string {
  const counts = `pulled ${res.issues} issues · ${res.comments} comments`;
  return res.removed > 0 ? `${counts} · removed ${res.removed}` : counts;
}

// syncRepo pulls the repo's tracker project into the hub issue store and sweeps
// the issues the tracker no longer returns, blocking for the length of both. A
// sync of that repo already running coalesces rather than pulling twice, so the
// caller waits on health instead of on this request.
export async function syncRepo(repo: string): Promise<SyncResult> {
  const res = await apiFetch(`/api/v1/repos/${encodeURIComponent(repo)}/sync`, {
    method: "POST",
  });
  if (!res.ok) {
    throw new Error(await errorMessage(res, "sync failed"));
  }
  if (res.status === 202) {
    return { status: "syncing" };
  }
  return { status: "pulled", response: await res.json() };
}

export interface DryRunResult {
  repo: string;
  repo_root: string;
  ticket: string;
}

export async function dryRun(repo: string): Promise<DryRunResult> {
  const res = await apiFetch(
    `/api/v1/repos/${encodeURIComponent(repo)}/dry-run`,
    { method: "POST" },
  );
  if (!res.ok) {
    throw new Error(await errorMessage(res, "dry-run failed"));
  }
  return res.json();
}
